// @meteorwallet/sdk/dist/meteor-common/src/models/external_actions/ExternalActionValidation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZO_MeteorSignTransactionAction_UrlQuery = exports.ZO_MeteorSignInAction_Combined = exports.ZO_MeteorSignInAction_AllMethods = exports.ZO_MeteorSignInAction_SelectedMethods = exports.ZO_MeteorSignInAction_Base = void 0;
const zod_1 = require("zod");
const ExternalActionTypes_1 = require("./ExternalActionTypes");
exports.ZO_MeteorSignInAction_Base = zod_1.z.object({
    contract_id: zod_1.z.string(),
    // callback_url: z.string(),
    // network: z.nativeEnum(ENearNetwork),
    public_key: zod_1.z.string(),
});
exports.ZO_MeteorSignInAction_SelectedMethods = exports.ZO_MeteorSignInAction_Base.merge(zod_1.z.object({
    type: zod_1.z.literal(ExternalActionTypes_1.EMeteorWalletSignInType.SELECTED_METHODS),
    methods: zod_1.z.array(zod_1.z.string()).min(1),
}));
exports.ZO_MeteorSignInAction_AllMethods = exports.ZO_MeteorSignInAction_Base.merge(zod_1.z.object({
    type: zod_1.z.literal(ExternalActionTypes_1.EMeteorWalletSignInType.ALL_METHODS),
    methods: zod_1.z.undefined().optional(),
}));
exports.ZO_MeteorSignInAction_Combined = exports.ZO_MeteorSignInAction_SelectedMethods.or(exports.ZO_MeteorSignInAction_AllMethods);
exports.ZO_MeteorSignTransactionAction_UrlQuery = zod_1.z.object({
    callback_url: zod_1.z.string().min(2),
    transactions: zod_1.z.string().min(2),
    meta: zod_1.z.string().optional(),
});
