// @nightlylabs/connect-near/lib/nightlyConnect/connect/connect-modal/src/style.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.modalStyle = void 0;
exports.modalStyle = `
  @import url('https://fonts.googleapis.com/css2?family=Prompt&display=swap');

  .nightlyConnect_overlay {
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    background: rgba(15, 15, 26, 0.4);
    backdrop-filter: blur(10px);
    z-index: 100;
  }

  .nightlyConnect_wrapper {
    box-sizing: border-box;
    max-width: 528px;
    background: #343551;
    border: 2px solid #707A8D;
    border-radius: 24px;
    padding: 32px;
    display: flex;
    flex-direction: column;
    margin: auto;
    position: relative;
    top: 50%;
    transform: translateY(-50%);
  }

  .nightlyConnect_topRow {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    height: 40px;
    width: 100%;
  }

  .nightlyConnect_logo {
    width: 146px;
    height: 36px;
  }

  .nightlyConnect_close {
    width: 40px;
    height: 40px;
    cursor: pointer;
  }

  .nightlyConnect_qrWrapper {
    background: #17182B;
    border-radius: 24px;
    padding: 12px;
    padding-bottom: 8px;
    margin-top: 28px;
    margin-bottom: 32px;
    width: 100%;
    box-sizing: border-box;
  }

  .nightlyConnect_qrImg {
    background-size: 440px 440px;
    max-width: 100%;
  }

  .nightlyConnect_bottomRow {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    width: 100%;
    box-sizing: border-box;
  }

  .nightlyConnect_desc {
    font-family: 'Prompt';
    font-style: normal;
    font-weight: 400;
    font-size: 16px;
    line-height: 22px;
    letter-spacing: 0.02em;
    color: #F7F7F7;
    text-align: start;
  }

  .nightlyConnect_copyButton {
    font-family: 'Prompt';
    font-style: normal;
    font-weight: 400;
    font-size: 14px;
    line-height: 18px;
    letter-spacing: 0.02em;
    color: #F7F7F7;
    background: #6067F9;
    border-radius: 6px;
    min-width: 88px;
    width: 88px;
    height: 32px;
    cursor: pointer;
    border-width: 0px;
    margin-left: 66px;
    transition: transform 300ms;
    padding: 0;
  }

  .nightlyConnect_copyButton:hover {
    transform: scale(1.2);
  }
`;
