// @nightlylabs/connect-near/lib/common/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RankingHomeCollectionEnum = exports.CollectionCategory = exports.NotificationType = exports.Device = exports.NETWORK_TICKERS = exports.NETWORK = void 0;
// ALL PROJECT
var NETWORK;
(function (NETWORK) {
    NETWORK["SOLANA"] = "SOLANA";
    NETWORK["NEAR"] = "NEAR";
    NETWORK["APTOS"] = "APTOS";
})(NETWORK = exports.NETWORK || (exports.NETWORK = {}));
exports.NETWORK_TICKERS = {
    [NETWORK.APTOS]: 'APT',
    [NETWORK.SOLANA]: 'SOL',
    [NETWORK.NEAR]: 'NEAR'
};
// MOBILE
var Device;
(function (Device) {
    Device["Apple"] = "Apple";
    Device["Android"] = "Android";
})(Device = exports.Device || (exports.Device = {}));
// TODO: Create intefaces for all Notifications Types
var NotificationType;
(function (NotificationType) {
    NotificationType["NIGHTLY_CONNECT_REQUEST"] = "NIGHTLY_CONNECT_REQUEST";
})(NotificationType = exports.NotificationType || (exports.NotificationType = {}));
// INDEXER
var CollectionCategory;
(function (CollectionCategory) {
    CollectionCategory["PFP"] = "PFP";
    CollectionCategory["ART"] = "ART";
    CollectionCategory["MUSIC"] = "MUSIC";
    CollectionCategory["METAVERSE"] = "METAVERSE";
    CollectionCategory["GAMES"] = "GAMES";
    CollectionCategory["SPORTS"] = "SPORTS";
    CollectionCategory["NSFW"] = "NSFW";
    CollectionCategory["EARNING"] = "EARNING";
    CollectionCategory["UTILITY"] = "UTILITY";
    CollectionCategory["VIRTUAL_WORLDS"] = "VIRTUAL_WORLDS";
    CollectionCategory["TRADING_CARDS"] = "TRADING_CARDS";
    CollectionCategory["OTHER"] = "OTHER";
})(CollectionCategory = exports.CollectionCategory || (exports.CollectionCategory = {}));
// MARKETPLACE
var RankingHomeCollectionEnum;
(function (RankingHomeCollectionEnum) {
    RankingHomeCollectionEnum["BLOCKCHAIN"] = "BLOCKCHAIN";
    RankingHomeCollectionEnum["TIME"] = "TIME";
    RankingHomeCollectionEnum["CATEGORY"] = "CATEGORY";
})(RankingHomeCollectionEnum = exports.RankingHomeCollectionEnum || (exports.RankingHomeCollectionEnum = {}));
