// @nightlylabs/connect-near/lib/nightlyConnect/connect/lib-near/src/persistedData.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearPersistedSessionAccountId = exports.clearPersistedSessionPublicKey = exports.clearPersistedSessionId = exports.setPersistedSessionAccountId = exports.setPersistedSessionPublicKey = exports.getPersistedSessionAccountId = exports.getPersistedSessionPublicKey = exports.getPersistedSessionId = exports.SESSION_ACCOUNTID_KEY = exports.SESSION_PUBLICKEY_KEY = exports.SESSION_ID_KEY = void 0;
const isomorphic_localstorage_1 = __importDefault(require("isomorphic-localstorage"));
const localStorage = (0, isomorphic_localstorage_1.default)('./.localstorage');
exports.SESSION_ID_KEY = 'nightly-id-near';
exports.SESSION_PUBLICKEY_KEY = 'nightly-connect-near-publickey';
exports.SESSION_ACCOUNTID_KEY = 'nightly-connect-near-accountid';
const getPersistedSessionId = () => localStorage.getItem(exports.SESSION_ID_KEY);
exports.getPersistedSessionId = getPersistedSessionId;
const getPersistedSessionPublicKey = () => localStorage.getItem(exports.SESSION_PUBLICKEY_KEY);
exports.getPersistedSessionPublicKey = getPersistedSessionPublicKey;
const getPersistedSessionAccountId = () => localStorage.getItem(exports.SESSION_ACCOUNTID_KEY);
exports.getPersistedSessionAccountId = getPersistedSessionAccountId;
const setPersistedSessionPublicKey = (publicKey) => {
    localStorage.setItem(exports.SESSION_PUBLICKEY_KEY, publicKey);
};
exports.setPersistedSessionPublicKey = setPersistedSessionPublicKey;
const setPersistedSessionAccountId = (accountId) => {
    localStorage.setItem(exports.SESSION_ACCOUNTID_KEY, accountId);
};
exports.setPersistedSessionAccountId = setPersistedSessionAccountId;
const clearPersistedSessionId = () => {
    localStorage.removeItem(exports.SESSION_ID_KEY);
};
exports.clearPersistedSessionId = clearPersistedSessionId;
const clearPersistedSessionPublicKey = () => {
    localStorage.removeItem(exports.SESSION_PUBLICKEY_KEY);
};
exports.clearPersistedSessionPublicKey = clearPersistedSessionPublicKey;
const clearPersistedSessionAccountId = () => {
    localStorage.removeItem(exports.SESSION_ACCOUNTID_KEY);
};
exports.clearPersistedSessionAccountId = clearPersistedSessionAccountId;
