// keypom-js/src/index.ts
export {
	addToBalance,
	withdrawBalance
} from "./lib/balances";
export {
	generateKeys,
	estimateRequiredDeposit,
	getStorageBase,
	ftTransferCall,
	nftTransferCall,
	hashPassword,
	exportedNearAPI as nearAPI,
} from "./lib/keypom-utils";
import { exportedNearAPI } from "./lib/keypom-utils";
export const {
	/** @group Utility */
	parseNearAmount, 
	/** @group Utility */
	formatNearAmount
} = exportedNearAPI.utils.format;
export {
	useKeypom,
	KeypomContextProvider,
} from './components/KeypomContext'
export {
	createDrop,
	deleteDrops,
} from "./lib/drops";
export {
	addKeys,
	deleteKeys
} from "./lib/keys";
export {
	claim,
} from "./lib/claims";
export {
	/** @group Utility */
	execute,
	initKeypom,
	getEnv,
	updateKeypomContractId,
	updateFunder
} from "./lib/keypom";
export * from "./lib/views";
export * from "./lib/types/drops";
export * from "./lib/types/fc";
export * from "./lib/types/ft";
export * from "./lib/types/general";
export * from "./lib/types/nft";
export * from "./lib/types/params";
export * from "./lib/types/simple";
export * from "./lib/types/protocol";
